src/appleseed.studio/mainwindow/rendering/renderingmanager.cpp

```diff
@@ -490,6 +495,105 @@ void RenderingManager::slot_camera_change_end()
 {
 }
 
+void RenderingManager::slot_material_dropped(const foundation::Vector2d& drop_pos, const std::string& material_name)
+{
+    if (!m_project->has_trace_context())
+    {
+        RENDERER_LOG_INFO("the scene must be rendering or must have been rendered at least once for drag&drop to be available.");
+        return;
+    }
+
+    DroppedMaterialMenu* menu = new DroppedMaterialMenu(drop_pos, material_name);
+
+    connect(menu,
+        SIGNAL(signal_apply_material(
+            const foundation::Vector2d&,
+            const std::string&,
+            renderer::ObjectInstance::Side)),
+        SLOT(slot_change_material(
+            const foundation::Vector2d&,
+            const std::string&,
+            renderer::ObjectInstance::Side)));
+
+    menu->exec(QCursor::pos());
+}
+
+namespace
+{
+    struct MaterialSlot
+    {
+        std::string name;
+        ObjectInstance::Side side;
+    };
+
+    class ApplyDroppedMaterial
+        : public RenderingManager::IScheduledAction
+    {
+    public:
+        explicit ApplyDroppedMaterial(
+            ObjectInstance& instance,
+            const std::vector<MaterialSlot>& material_slots,
+            const std::string& material_name)
+            : m_instance(instance),
+            m_material_slots(material_slots),
+            m_material_name(material_name)
+        {
+        }
+
+        void operator()(Project& project) override
+        {
+            for (const_each<std::vector<MaterialSlot>> slot = m_material_slots; slot; ++slot)
+            {
+                m_instance.assign_material(slot->name.c_str(), slot->side, m_material_name.c_str());
+            }
+        }
+
+    private:
+        ObjectInstance& m_instance;
+        const std::vector<MaterialSlot> m_material_slots;
+        const std::string m_material_name;
+    };
+}
+
+void RenderingManager::slot_change_material(
+    const foundation::Vector2d& drop_pos,
+    const std::string& material_name,
+    renderer::ObjectInstance::Side side)
+{
+    const ScenePicker scene_picker(*m_project);
+    const ScenePicker::PickingResult result = scene_picker.pick(drop_pos);
+    renderer::ObjectInstance* instance = const_cast<renderer::ObjectInstance*>(result.m_object_instance);
```

Comment:
I know we discussed that earlier, but I would prefer that you make `ScenePicker::PickingResult` members non-const instead.

Reply:
ScenePicker retreive m_object_instance from shadingpoint.cpp L164
`// Retrieve the object instance.
    m_object_instance = m_assembly->object_instances().get_by_index(m_object_instance_index);
    assert(m_object_instance);`
and I didnt find a way to make this line retreive a non-const object_instance

Reply:
I see. I guess we need to turn `ShadingPoint::m_object_instance` into a mutable pointer then.